lib/xenops_server_simulator.ml

```diff
@@ -142,7 +145,11 @@ let get_domain_action_request_nolock vm () =
 let destroy_nolock vm () =
   debug "Domain.destroy vm=%s" vm.Vm.id ;
   (* Idempotent *)
-  if DB.exists vm.Vm.id then DB.delete vm.Vm.id
+  if DB.exists vm.Vm.id then
+    let d = DB.read_exn vm.Vm.id in
+    (* Preserve the domain state if it has suspended *)
```

Comment:
question: why is it correct to keep suspended VMs in the xenopsd database, when we want to destroy the VM...? I don't see how this ties into preserving domain state

Reply:
The `VM.get_state` function in xenops_server_xen.ml (non-simulator) looks at the underlying domain to determine the power state. The simulator doesn't explicitly model domains and conflates the concepts a bit by treating VMs that are in the DB as "running" and "halted" otherwise.

The purpose of the `destroy` function is mainly to remove any resources that were in use for the VM, such as the domain. In the simulator, there is no such thing, so that part is a no-op. The `destroy` in the real backend has similar logic for preserving the runtime state in the DB for suspended VMs: https://github.com/xapi-project/xenopsd/blob/5176b5b372fb2a67be9d5477d80d6080abbe52b8/xc/xenops_server_xen.ml#L1558-L1575.  This is to allow xapi to retrieve and store the state (in the field `VM.last_booted_record`) after the suspend call. The change in this patch makes the simulator behave a little more like the real backend.

Reply:
tldr, we shouldn't delete the VM from xenopsd DB if it's shutdown, because xapi needs to look up `last_booted_record` (i.e. destroy shouldn't _always_ destroy)